### PR TITLE
add tcnp change detection

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards Node Pools. Preparing updates. The usage of the change detection service added here will be wired in follow up PRs.

### CODE REVIEW

File: service/controller/clusterapi/v29/controllercontext/status_types.go [hunk 1]
```diff
@@ -113,3 +114,8 @@ type ContextStatusTenantClusterTCNPASG struct {
 	MinSize         int
 	Name            string
 }
+
+type ContextStatusTenantClusterTCNPWorkerInstance struct {
+	DockerVolumeSizeGB string
```
Comment (MarcelMue) [inline service/controller/clusterapi/v29/controllercontext/status_types.go]: This then also needs to be filled in a followup.

File: service/controller/clusterapi/v29/changedetection/tcnp.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+//     The tenant cluster's scaling max changes.
+//     The tenant cluster's scaling min changes.
+//
+func (t *TCNP) ShouldScale(ctx context.Context, md v1alpha1.MachineDeployment) (bool, error) {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	if !cc.Status.TenantCluster.TCNP.ASG.IsEmpty() && cc.Status.TenantCluster.TCNP.ASG.MaxSize != key.MachineDeploymentScalingMax(md) {
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/changedetection/tcnp.go]: I'd propose to potentially break these down a bit for easier readability:
```go
asgExists := !cc.Status.TenantCluster.TCNP.ASG.IsEmpty()
asgMaxNumberOfNodesDiffer := cc.Status.TenantCluster.TCNP.ASG.MaxSize != key.MachineDeploymentScalingMax(md)
asgMinNumberOfNodesDiffer := cc.Status.TenantCluster.TCNP.ASG.MinSize != key.MachineDeploymentScalingMin(md)

if asgExists && asgMaxNumberOfNodesDiffer {
    // ....
}

if asgExists && asgMinNumberOfNodesDiffer {
    // ....
}
```

This way it's not necessary to list these in above function comment which might get forgotten to update when later adding more rules here. That kind of explicit variable naming makes the code basically the same as comment and therefore documents itself quite nicely.

File: service/controller/clusterapi/v29/changedetection/tcnp.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+//     The worker node's instance type changes.
+//     The tenant cluster's version changes.
+//
+func (t *TCNP) ShouldUpdate(ctx context.Context, md v1alpha1.MachineDeployment) (bool, error) {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	if cc.Status.TenantCluster.TCNP.WorkerInstance.DockerVolumeSizeGB != key.MachineDeploymentDockerVolumeSizeGB(md) {
```
Comment (tuommaki) [inline service/controller/clusterapi/v29/changedetection/tcnp.go]: Potentially the same here as above.

File: service/controller/clusterapi/v29/controllercontext/status_types.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/controllercontext/status_types.go]: Yes, I am about to create the `tcnpoutputs` resource.

File: service/controller/clusterapi/v29/changedetection/tcnp.go (hunk 2 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/changedetection/tcnp.go]: Good shout. Done that.

**Pull request merged**